components/kyma-environment-broker/internal/process/provisioning/apply_kyma_step.go

```diff
@@ -0,0 +1,108 @@
+package provisioning
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/process/steps"
+
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/process"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/storage"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	yamlutil "k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	kymaResourceNamespace = "kcp-system"
+)
+
+type ApplyKymaStep struct {
+	operationManager *process.OperationManager
+}
+
+var _ process.Step = &ApplyKymaStep{}
+var kymaGVK = schema.GroupVersionKind{Group: "operator.kyma-project.io", Version: "v1alpha1", Kind: "Kyma"}
+
+func NewApplyKymaStep(os storage.Operations) *ApplyKymaStep {
+	return &ApplyKymaStep{operationManager: process.NewOperationManager(os)}
+}
+
+func (a *ApplyKymaStep) Name() string {
+	return "Apply_Kyma"
+}
+
+func (a *ApplyKymaStep) Run(operation internal.Operation, logger logrus.FieldLogger) (internal.Operation, time.Duration, error) {
+	if operation.K8sClient == nil {
+		return a.operationManager.OperationFailed(operation, "operation does not contain initialized k8s client", nil, logger)
+	}
+	k8s := operation.K8sClient
+
+	template, err := a.createUnstructuredKyma(operation)
+	if err != nil {
+		return a.operationManager.OperationFailed(operation, "unable to create a kyma template", err, logger)
+
+	}
+	a.addLabelsAndName(operation, template)
+
+	var existingKyma unstructured.Unstructured
+	existingKyma.SetGroupVersionKind(kymaGVK)
+	err = k8s.Get(context.Background(), client.ObjectKey{
+		Namespace: template.GetNamespace(),
+		Name:      template.GetName(),
+	}, &existingKyma)
+	switch {
+	case err == nil:
+		logger.Infof("Kyma resource already exists, updating")
+		a.addLabelsAndName(operation, &existingKyma)
+		err = k8s.Update(context.Background(), &existingKyma)
```

Comment:
nit: small performance optimization, I am considering doing diff between `template` and `existingKyma` in https://github.com/kyma-project/control-plane/pull/2216 (not there yet), we can prevent API call if there is no diff..

Reply:
KEB is not the only one component who manage this resource. The spec could be changed. Also somebody else can modify labels. I can imagine that we can check if labels where modified, then decide if we need the update, but it is a rare case in my opinion

Reply:
I think we agree, KEB doesn't need to call `Update` unless there is a diff between `template` and `existingKyma`. But can be considered an optimization candidate for the future.